src/dowel/csv_output.py

```diff
@@ -2,59 +2,66 @@
+from dowel.tabular import Tabular
 from dowel.utils import colorize
 
 
 class CsvOutput(FileOutput):
     """CSV file output for logger.
 
     :param file_name: The file this output should log to.
+    :param keys_accepted: Regex for which keys this output should accept.
     """
 
-    def __init__(self, file_name):
-        super().__init__(file_name)
+    def __init__(self, file_name, keys_accepted=r'^\S+$'):
+        super().__init__(file_name, keys_accepted=keys_accepted)
         self._writer = None
         self._fieldnames = None
         self._warned_once = set()
         self._disable_warnings = False
+        self.tabular = Tabular()
 
     @property
     def types_accepted(self):
-        """Accept TabularInput objects only."""
-        return (TabularInput, )
-
-    def record(self, data, prefix=''):
-        """Log tabular data to CSV."""
-        if isinstance(data, TabularInput):
-            to_csv = data.as_primitive_dict
-
-            if not to_csv.keys() and not self._writer:
-                return
-
-            if not self._writer:
-                self._fieldnames = set(to_csv.keys())
-                self._writer = csv.DictWriter(
-                    self._log_file,
-                    fieldnames=self._fieldnames,
-                    extrasaction='ignore')
-                self._writer.writeheader()
-
-            if to_csv.keys() != self._fieldnames:
-                self._warn('Inconsistent TabularInput keys detected. '
-                           'CsvOutput keys: {}. '
-                           'TabularInput keys: {}. '
-                           'Did you change key sets after your first '
-                           'logger.log(TabularInput)?'.format(
-                               set(self._fieldnames), set(to_csv.keys())))
-
-            self._writer.writerow(to_csv)
-
-            for k in to_csv.keys():
-                data.mark(k)
-        else:
-            raise ValueError('Unacceptable type.')
+        """Accept str and scalar objects."""
+        return (str, ) + np.ScalarType
+
+    def record(self, key, value, prefix=''):
+        """Log data to a csv file."""
+        self.tabular.record(key, value)
+
+    def dump(self, step=None):
+        """Flush data to log file."""
+        if self.tabular.empty:
+            return
+
+        to_csv = self.tabular.as_primitive_dict
+
+        if not to_csv.keys() and not self._writer:
+            return
+
+        if not self._writer:
+            self._fieldnames = set(to_csv.keys())
+            self._writer = csv.DictWriter(self._log_file,
+                                          fieldnames=self._fieldnames,
+                                          extrasaction='ignore')
+            self._writer.writeheader()
+
+        if to_csv.keys() != self._fieldnames:
+            self._warn('Inconsistent Tabular keys detected. '
+                       'CsvOutput keys: {}. '
+                       'Tabular keys: {}. '
```

Comment:
your user now has no idea that a Tabular is, so this message needs to be updated.